File: frontend/src/utils/mqtt.ts
import { connect, MqttClient } from 'mqtt';

interface MqttConfig {
    brokerUrl: string;
    clientId: string;
    username?: string;
    password?: string;
}

export class MqttManager {
    private client: MqttClient | null = null;
    private subscriptions: Map<string, ((message: any) => void)[]> = new Map();

    async connect(config: MqttConfig): Promise<void> {
        return new Promise((resolve, reject) => {
            this.client = connect(config.brokerUrl, {
                clientId: config.clientId,
                username: config.username,
                password: config.password,
                clean: true,
            });

            this.client.on('connect', () => {
                console.log('Connected to MQTT broker');
                resolve();
            });

            this.client.on('error', (error) => {
                console.error('MQTT connection error:', error);
                reject(error);
            });

            this.client.on('message', (topic, message) => {
                const handlers = this.subscriptions.get(topic);
                if (handlers) {
                    const payload = JSON.parse(message.toString());
                    handlers.forEach(handler => handler(payload));
                }
            });
        });
    }

    subscribe(topic: string, handler: (message: any) => void): void {
        if (!this.client) {
            throw new Error('MQTT client not connected');
        }

        const handlers = this.subscriptions.get(topic) || [];
        if (handlers.length === 0) {
            this.client.subscribe(topic);
        }
        handlers.push(handler);
        this.subscriptions.set(topic, handlers);
    }

    unsubscribe(topic: string, handler: (message: any) => void): void {
        if (!this.client) return;

        const handlers = this.subscriptions.get(topic);
        if (handlers) {
            const index = handlers.indexOf(handler);
            if (index > -1) {
                handlers.splice(index, 1);
            }
            if (handlers.length === 0) {
                this.client.unsubscribe(topic);
                this.subscriptions.delete(topic);
            }
        }
    }

    publish(topic: string, message: any): void {
        if (!this.client) {
            throw new Error('MQTT client not connected');
        }

        this.client.publish(topic, JSON.stringify(message));
    }

    disconnect(): void {
        if (this.client) {
            this.client.end();
            this.client = null;
            this.subscriptions.clear();
        }
    }
}

export const mqttManager = new MqttManager();

File: backend/pkg/collaboration/mqtt.go
package collaboration

import (
	"encoding/json"
	"fmt"
	"sync"
	"time"

	"github.com/HardMax71/syncwrite/backend/pkg/utils"
	mqtt "github.com/eclipse/paho.mqtt.golang"
	"go.uber.org/zap"
)

type MQTTClient struct {
	client   mqtt.Client
	logger   *zap.Logger
	handlers map[string]func([]byte)
	mutex    sync.RWMutex
}

func NewMQTTClient(brokerURL string) (*MQTTClient, error) {
	logger := utils.Logger()

	opts := mqtt.NewClientOptions().
		AddBroker(brokerURL).
		SetClientID(fmt.Sprintf("syncwrite-server-%d", time.Now().UnixNano())).
		SetCleanSession(true).
		SetAutoReconnect(true).
		SetOnConnectHandler(func(client mqtt.Client) {
			logger.Info("Connected to MQTT broker")
		}).
		SetConnectionLostHandler(func(client mqtt.Client, err error) {
			logger.Error("Lost connection to MQTT broker", zap.Error(err))
		})

	client := mqtt.NewClient(opts)
	if token := client.Connect(); token.Wait() && token.Error() != nil {
		return nil, fmt.Errorf("error connecting to MQTT broker: %w", token.Error())
	}

	return &MQTTClient{
		client:   client,
		logger:   logger,
		handlers: make(map[string]func([]byte)),
	}, nil
}

func (m *MQTTClient) Subscribe(topic string, handler func([]byte)) error {
	m.mutex.Lock()
	m.handlers[topic] = handler
	m.mutex.Unlock()

	token := m.client.Subscribe(topic, 1, func(client mqtt.Client, msg mqtt.Message) {
		m.mutex.RLock()
		if handler, exists := m.handlers[msg.Topic()]; exists {
			handler(msg.Payload())
		}
		m.mutex.RUnlock()
	})

	token.Wait()
	return token.Error()
}

func (m *MQTTClient) Unsubscribe(topic string) error {
	m.mutex.Lock()
	delete(m.handlers, topic)
	m.mutex.Unlock()

	token := m.client.Unsubscribe(topic)
	token.Wait()
	return token.Error()
}

func (m *MQTTClient) Publish(topic string, payload interface{}) error {
	data, err := json.Marshal(payload)
	if err != nil {
		return fmt.Errorf("error marshaling payload: %w", err)
	}

	token := m.client.Publish(topic, 1, false, data)
	token.Wait()
	return token.Error()
}

func (m *MQTTClient) Close() {
	m.client.Disconnect(250)
}

func GetDocumentTopic(documentID string) string {
	return fmt.Sprintf("documents/%s/changes", documentID)
}

func GetPresenceTopic(documentID string) string {
	return fmt.Sprintf("documents/%s/presence", documentID)
}
